### PR TITLE
[PWGHF] Adding matching of candidates with kinked daughter tracks to Charm Resonances workflow

### PR DESCRIPTION
**Pull request by Luca610**
This PR implement the recent changes to RecoDecay.h to match MC candidates whose daughter tracks decay with kink topologies. Optional histograms are added to study the shape of the invariant mass distribution of these decays.
The MCflag naming convention has been changed to accomodate for new matching flags in the future.

### CODE REVIEW

File: PWGHF/D2H/TableProducer/dataCreatorCharmResoReduced.cxx [hunk 1]
```diff
@@ -245,15 +253,35 @@ struct HfDataCreatorCharmResoReduced {
...
     registry.add("hMCGenOrigin", "Origin of Generated particles", {HistType::kTH1F, {{3, -0.5, 2.5}}});
     registry.add("hMCOriginCounterWrongDecay", "Origin of Generated particles in Wrong decay", {HistType::kTH1F, {{3, -0.5, 2.5}}});
 
+    if (doMcRecQa) {
+      registry.add("hMassVsPtK0Matched", "K0s candidates Matched ;#it{p}_{T} (GeV/#it{c});inv. mass (GeV/#it{c}^{2});entries", {HistType::kTH2F, {axisPt, axisMassKzero}});
+      registry.add("hMassVsPtD0Matched", "D0 candidates Matched ;#it{p}_{T} (GeV/#it{c});inv. mass (GeV/#it{c}^{2});entries", {HistType::kTH2F, {axisPt, axisMassDplus}});
+      registry.add("hMassVsPtDstarMatched", "Dstar candidates Matched ;#it{p}_{T} (GeV/#it{c});inv. mass (GeV/#it{c}^{2});entries", {HistType::kTH2F, {axisPt, axisMassDstar}});
+      registry.add("hMassVsPtDplusMatched", "Dplus candidates Matched ;#it{p}_{T} (GeV/#it{c});inv. mass (GeV/#it{c}^{2});entries", {HistType::kTH2F, {axisPt, axisMassDplus}});
+      registry.add("hMassVsPtDs1Matched", "Ds1 candidates Matched ;#it{p}_{T} (GeV/#it{c});inv. mass (GeV/#it{c}^{2});entries", {HistType::kTH2F, {axisPt, {400, 0.49, 0.89}}});
```
Comment (fgrosa) [inline PWGHF/D2H/TableProducer/dataCreatorCharmResoReduced.cxx:270]: define an `AxisSpec` for the delta mass (`{400, 0.49, 0.89}` is repeated many times)

File: PWGHF/D2H/TableProducer/dataCreatorCharmResoReduced.cxx [hunk 2]
```diff
@@ -517,138 +545,129 @@ struct HfDataCreatorCharmResoReduced {
...
+          flag = sign * DecayTypeMc::Ds1ToDStarK0ToD0PiK0s;
+        } else if (indexRecReso > -1 && nPiToMuReso >= 1 && nKaToPiReso == 0) {
+          flag = sign * DecayTypeMc::Ds1ToDStarK0ToD0PiK0sOneMu;
         }
       }
 
       // Ds1+ not matched: we check if it is partially reco
       if (indexRecReso < 0) {
-        indexRecResoPartReco = RecoDecay::getMatchedMCRec<false, true, true>(particlesMc, std::array{vecDaughtersReso[0], vecDaughtersReso[1], vecDaughtersReso[2], vecDaughtersReso[3], vecDaughtersReso[4]}, Pdg::kDS1, std::array{+kPiPlus, -kKPlus, +kPiPlus, +kPiPlus, -kPiPlus}, true, &sign, 3);
+        indexRecResoPartReco = RecoDecay::getMatchedMCRec<false, true, true, true>(particlesMc, std::array{vecDaughtersReso[0], vecDaughtersReso[1], vecDaughtersReso[2], vecDaughtersReso[3], vecDaughtersReso[4]}, Pdg::kDS1, std::array{+kPiPlus, -kKPlus, +kPiPlus, +kPiPlus, -kPiPlus}, true, &sign, 3);
```
Comment (fgrosa) [inline PWGHF/D2H/TableProducer/dataCreatorCharmResoReduced.cxx:572]: why is the check on muons enabled here, but we do not count them?

File: PWGHF/D2H/TableProducer/dataCreatorCharmResoReduced.cxx [hunk 3]
```diff
@@ -517,138 +545,129 @@ struct HfDataCreatorCharmResoReduced {
...
         indexRecDplus = RecoDecay::getMatchedMCRec(particlesMc, std::array{vecDaughtersReso[0], vecDaughtersReso[1], vecDaughtersReso[2]}, Pdg::kDPlus, std::array{+kPiPlus, -kKPlus, +kPiPlus}, true, &signDPlus, 2);
         if (indexRecResoPartReco > -1) { // we look for decays of D* or D0 with more daughters
           if (indexRecDstar < 0 && indexRecK0 > -1) {
-            auto indexRecDstarPartReco = RecoDecay::getMatchedMCRec<false, false, true>(particlesMc, std::array{vecDaughtersReso[0], vecDaughtersReso[1], vecDaughtersReso[2]}, Pdg::kDStar, std::array{-kKPlus, +kPiPlus, +kPiPlus}, true, &signDStar, 3);
+            auto indexRecDstarPartReco = RecoDecay::getMatchedMCRec<false, false, true, true>(particlesMc, std::array{vecDaughtersReso[0], vecDaughtersReso[1], vecDaughtersReso[2]}, Pdg::kDStar, std::array{-kKPlus, +kPiPlus, +kPiPlus}, true, &signDStar, 3);
```
Comment (fgrosa) [inline PWGHF/D2H/TableProducer/dataCreatorCharmResoReduced.cxx:576]: same here (no counters for muons)

File: PWGHF/D2H/TableProducer/dataCreatorCharmResoReduced.cxx [hunk 4]
```diff
@@ -517,138 +545,129 @@ struct HfDataCreatorCharmResoReduced {
...
             if (indexRecDstarPartReco > -1) {
               if (indexRecDplus > -1) { // Ds1 -> D* K0s -> D+  π0 K0s
-                flag = sign * BIT(DecayTypeMc::Ds1ToDStarK0ToDPlusPi0K0s);
+                flag = sign * DecayTypeMc::Ds1ToDStarK0ToDPlusPi0K0s;
               } else {
-                auto indexRecDzeroPartReco = RecoDecay::getMatchedMCRec<false, false, true>(particlesMc, std::array{vecDaughtersReso[0], vecDaughtersReso[1]}, Pdg::kD0, std::array{+kPiPlus, -kKPlus}, true, &signD0, 2);
+                auto indexRecDzeroPartReco = RecoDecay::getMatchedMCRec<false, false, true, true>(particlesMc, std::array{vecDaughtersReso[0], vecDaughtersReso[1]}, Pdg::kD0, std::array{+kPiPlus, -kKPlus}, true, &signD0, 2);
```
Comment (fgrosa) [inline PWGHF/D2H/TableProducer/dataCreatorCharmResoReduced.cxx:581]: same here

File: PWGHF/D2H/TableProducer/dataCreatorCharmResoReduced.cxx [hunk 5]
```diff
@@ -517,138 +545,129 @@ struct HfDataCreatorCharmResoReduced {
...
-                  flag = sign * BIT(DecayTypeMc::Ds1ToDStarK0ToD0PiK0sPart);
+                  flag = sign * DecayTypeMc::Ds1ToDStarK0ToD0PiK0sPart;
                 }
               }
             }
           }
         } else { // we look for D* not matched, but all the other ones yes, we check if we only lost the soft pion
           if (indexRecD0 > -1 && indexRecK0 > -1 && indexRecDstar < 0) {
-            indexRecResoPartReco = RecoDecay::getMatchedMCRec<false, true, true>(particlesMc, std::array{vecDaughtersReso[0], vecDaughtersReso[1], vecDaughtersReso[3], vecDaughtersReso[4]}, Pdg::kDS1, std::array{+kPiPlus, -kKPlus, +kPiPlus, -kPiPlus}, true, &sign, 3);
+            indexRecResoPartReco = RecoDecay::getMatchedMCRec<false, true, true, true>(particlesMc, std::array{vecDaughtersReso[0], vecDaughtersReso[1], vecDaughtersReso[3], vecDaughtersReso[4]}, Pdg::kDS1, std::array{+kPiPlus, -kKPlus, +kPiPlus, -kPiPlus}, true, &sign, 3);
```
Comment (fgrosa) [inline PWGHF/D2H/TableProducer/dataCreatorCharmResoReduced.cxx:590]: same here

**Review by fgrosa — CHANGES_REQUESTED** [review]
Hi @Luca610 I have a few minor comments, then the PR looks good to me, thanks!

File: PWGHF/D2H/TableProducer/dataCreatorCharmResoReduced.cxx (hunk 2 above)
Comment (Luca610) [reply PWGHF/D2H/TableProducer/dataCreatorCharmResoReduced.cxx:572]: In Reco Decay the counter is optional, it is needed only if you want to discriminate between candidates matched with one muon or not. By not providing the counter and not applying andy selection on the number of muons I am just keeping the candidates matched with muons in the same group of the ones matched without any of them... I am doing it mainly for the correlated backgrounds to keep the number of candidates classes low

**Review by fgrosa — APPROVED** [review]

**Pull request merged**